openvdb_points/tools/IndexFilter.h

```diff
@@ -0,0 +1,495 @@
+
+    if (n >= m) {
+        values.reserve(n);
+        for (int i = 0; i < int(m); i++)    values.push_back(IntType(i));
+        return;
+    }
+
+    // reserve and fill vector with ascending indices
+    values.reserve(m);
+    for (int i = 0; i < int(m); i++)    values.push_back(i);
+
+    // shuffle indices using random generator
+    RandGen<RandGenT, IntType> randGen(seed, m);
+    std::random_shuffle(values.begin(), values.end(), randGen);
+
+    // resize the container to n elements (does not reduce capacity)
+    values.resize(n);
+
+    // sort the subset of the indices vector that will be used
+    std::sort(values.begin(), values.end());
+}
+
+
+} // namespace index_filter_internal
+
+
+/// Index filtering on multiple group membership for inclusion and exclusion
+class MultiGroupFilter
+{
+public:
+    typedef std::vector<Name> NameVector;
+    typedef boost::ptr_vector<GroupHandle> HandleVector;
+
+    struct Data
+    {
+        Data(   const NameVector& _include,
+                const NameVector& _exclude)
+            : include(_include)
+            , exclude(_exclude) { }
+        const NameVector include;
+        const NameVector exclude;
+    };
+
+    MultiGroupFilter(   const HandleVector& includeHandles,
+                        const HandleVector& excludeHandles)
+        : mIncludeHandles(includeHandles)
+        , mExcludeHandles(excludeHandles) { }
+
+    template <typename LeafT>
+    static MultiGroupFilter create(const LeafT& leaf, const Data& data) {
+        HandleVector include;
+        HandleVector exclude;
+        for (NameVector::const_iterator  it = data.include.begin(),
+                                                itEnd = data.include.end(); it != itEnd; ++it) {
+            if (!leaf.attributeSet().descriptor().hasGroup(*it))    continue;
+            include.push_back(new GroupHandle(leaf.groupHandle(*it)));
+        }
+        for (NameVector::const_iterator  it = data.exclude.begin(),
+                                                itEnd = data.exclude.end(); it != itEnd; ++it) {
+            if (!leaf.attributeSet().descriptor().hasGroup(*it))    continue;
+            exclude.push_back(new GroupHandle(leaf.groupHandle(*it)));
+        }
+        return MultiGroupFilter(include, exclude);
+    }
+
+    template <typename IterT>
+    bool valid(const IterT& iter) const {
+        for (HandleVector::const_iterator   it = mIncludeHandles.begin(),
+                                            itEnd = mIncludeHandles.end(); it != itEnd; ++it) {
+            if (!it->get(*iter))    return false;
+        }
+        for (HandleVector::const_iterator   it = mExcludeHandles.begin(),
+                                            itEnd = mExcludeHandles.end(); it != itEnd; ++it) {
+            if (it->get(*iter))     return false;
+        }
+        return true;
+    }
+
+private:
+    HandleVector mIncludeHandles;
+    HandleVector mExcludeHandles;
+}; // class MultiGroupFilter
+
+
+struct LeafMapData
```

Comment:
Only seems to be used by RandomLeafFilter - can we move this into the RandomLeafFilter scope?